File: backend/app/schemas/config_version.py
from __future__ import annotations

import json
from typing import Any, Literal

from pydantic import BaseModel, field_validator


class ConfigVersionCreate(BaseModel):
    yaml_content: str
    source_tag: Literal["user", "ai_suggestion", "system"]
    source_detail: str | None = None


class ConfigVersionResponse(BaseModel):
    id: str
    project_id: str
    version_number: int
    yaml_hash: str
    diff_from_prev: dict[str, Any] | None
    source_tag: str
    source_detail: str | None
    created_at: str

    model_config = {"from_attributes": True}

    @field_validator("diff_from_prev", mode="before")
    @classmethod
    def parse_diff_json(cls, v: str | dict | None) -> dict[str, Any] | None:
        if v is None:
            return None
        if isinstance(v, str):
            return json.loads(v)  # type: ignore[no-any-return]
        return v


class ConfigVersionListResponse(BaseModel):
    items: list[ConfigVersionResponse]
    total: int
    limit: int
    offset: int


class ConfigDiffResponse(BaseModel):
    version_a_id: str
    version_b_id: str
    diff: dict[str, Any]


class ConfigValidationResponse(BaseModel):
    is_valid: bool
    errors: list[str]

File: frontend/src/types/config-version.ts
export interface ConfigVersionSummary {
  readonly id: string;
  readonly projectId: string;
  readonly versionNumber: number;
  readonly yamlHash: string;
  readonly diffFromPrev: unknown | null;
  readonly sourceTag: string;
  readonly sourceDetail: string | null;
  readonly createdAt: string;
}

export interface ConfigVersionList {
  readonly items: ReadonlyArray<ConfigVersionSummary>;
  readonly total: number;
  readonly limit: number;
  readonly offset: number;
}

export interface ConfigValidationResult {
  readonly isValid: boolean;
  readonly errors: ReadonlyArray<string>;
}
